File: Python/knightsearch.py
def main():
	to_find = "ICPCASIASG" # String to spell out on chessboard
	dirs = [(2,1),(2,-1),(-2,1),(-2,-1),(1,2),(1,-2),(-1,2),(-1,-2)]
	
	# Take inputs
	n = int(input())
	line = input()
	# Create structures used in search
	graph = []
	q = []
	
	# Construct graph and find starting points ("I"s)
	for i in range(n):
		graph.append(line[n*i:n*i+n])
		for j in range(n):
			if graph[i][j] == 'I':
				q.append((i,j,0))

	found = False
	# Perform a DFS from each starting position
	while q:
		y,x,curr = q.pop()

		# Check if at the end of the string
		if curr == 9:
			found = True
			break

		for i,j in dirs:
			# Make sure new point is in bounds and the letter im looking for
			if 0 <= y+i < n and 0 <= x+j < n:
				if graph[y+i][x+j] == to_find[curr+1]:
					q.append((y+i,x+j,curr+1))

	print("YES" if found else "NO")

main()

File: Java/KnightSearch.java
import java.util.Scanner;
import java.util.Stack;

public class KnightSearch {
	static int length;
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		Stack<Tuple> stack = new Stack<Tuple>();
		String search = "ICPCASIASG";

		length = in.nextInt();
		String board = in.next();

		Tuple[] dirs = {
			new Tuple(2,1,1),
			new Tuple(2,-1,1),
			new Tuple(-2,1,1),
			new Tuple(-2,-1,1),
			new Tuple(1,2,1),
			new Tuple(1,-2,1),
			new Tuple(-1,2,1),
			new Tuple(-1,-2,1)
		};

		for (int i = 0; i < length; i++) {
			for (int j = 0; j < length; j++) {
				if (board.charAt(i*length+j) == 'I') stack.push(new Tuple(j,i,0));
			}
		}

		while (!stack.isEmpty()) {
			Tuple curr = stack.pop();
			if (curr.i == 9) {
				System.out.println("YES");
				return;
			}

			for (Tuple d : dirs) {
				Tuple n = curr.add(d);

				if (n.onBoard() && board.charAt(n.y*length+n.x) == search.charAt(n.i)) {
					stack.push(n);
				}
			}
		}

		System.out.println("NO");
	}

	static class Tuple {
		int x;
		int y;
		int i;

		public Tuple(int x, int y, int i) {
			this.x = x;
			this.y = y;
			this.i = i;
		}

		public Tuple add(Tuple t) {
			return new Tuple(this.x + t.x, this.y + t.y, this.i + t.i);
		}

		public boolean onBoard() {
			return (this.x >= 0 && this.x < length
					  && this.y >= 0 && this.y < length);
		}
	}
}